source/language_service/src/wrapper_refactor.rs

```diff
@@ -0,0 +1,297 @@
+    encoding: Encoding,
+) -> Vec<CodeAction> {
+    let mut code_actions = Vec::new();
+    let user_unit = compilation.user_unit();
+    let package = &user_unit.package;
+    let source_map = &user_unit.sources;
+    let source = source_map
+        .find_by_name(source_name)
+        .expect("source should exist");
+    let source_span = compilation.package_span_of_source(source_name);
+
+    for (_, item) in package.items.iter() {
+        if !source_span.contains(item.span.lo) || span.intersection(&item.span).is_none() {
+            continue;
+        }
+        if let ItemKind::Callable(decl) = &item.kind {
+            if decl.kind != CallableKind::Operation || decl.input.ty == Ty::UNIT {
+                continue; // only operations with non-empty params
+            }
+
+            // Determine indentation using source-local offset (package offset minus source base).
+            let local_lo = item.span.lo - source.offset;
+            let indent = line_indentation(&source.contents, local_lo);
+            let body_indent = if indent.contains('\t') {
+                format!("{indent}\t")
+            } else {
+                format!("{indent}    ")
+            };
+
+            let original_name = decl.name.name.as_ref();
+            let wrapper_name = generate_unique_wrapper_name(package, original_name);
+
+            let (decl_lines, call_args) = build_param_decls_and_call_args(&decl.input);
+
+            let call_args_joined = if call_args.is_empty() {
+                String::new()
+            } else {
+                call_args.join(", ")
+            };
+
+            let return_ty = decl.output.display();
+            let return_is_unit = decl.output == Ty::UNIT;
+
+            let call_line = if return_is_unit {
+                format!("{body_indent}{original_name}({call_args_joined});")
+            } else {
+                format!("{body_indent}return {original_name}({call_args_joined});")
+            };
+
+            let mut body_lines = Vec::new();
+            if !decl_lines.is_empty() {
+                body_lines.push(format!(
+                    "{body_indent}// TODO: Fill out the values for the parameters"
+                ));
+                body_lines.extend(decl_lines.iter().map(|decl| format!("{body_indent}{decl}")));
+                body_lines.push(String::new()); // blank line
+            }
+            body_lines.push(format!("{body_indent}// Call original operation"));
+            body_lines.push(call_line);
+
+            // We intentionally do NOT prefix the first line with `indent` because the insertion point
+            // inherits the existing line's leading whitespace. We DO append `{indent}` after the blank line
+            // so that the original operation keeps its indentation after the inserted block.
+            let wrapper_text = format!(
+                "operation {wrapper_name}() : {return_ty} {{\n{}\n{indent}}}\n\n{indent}",
+                &body_lines.join("\n")
```

Comment:
How do you respect the chosen line ending (`\r\n` or `\n`) of the existing document?

If this is difficult, I wonder if you can return this as multiple text edits (one for each line) and let VS Code stitch them together

Reply:
I added some logic to detect the line ending used in the source and use that instead of just assuming '\n'.